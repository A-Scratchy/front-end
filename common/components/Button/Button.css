.Button {
  align-items: center;
  border: 1px solid transparent;
  border-radius: 0.25rem;
  cursor: pointer;
  display: flex;
  font-size: 1rem;
  justify-content: center;
  line-height: 1.25;
  min-width: 200px;
  padding: 0.75rem;
  transition: 0.1s linear background-color;
  text-decoration: none;
}

.Button:active,
.Button:focus {
  outline-color: #000000;
  outline-offset: -2px;
  outline-style: solid;
  outline-width: 2px;
}

.fullWidth {
  width: 100%;
}

.primary {
  background-color: var(--primary);
  border-color: var(--primaryDark);
  color: #ffffff;
  outline-color: var(--primaryDark);
}

.primary:hover {
  background-color: var(--primaryLight);
}

.primary:active:not:hover {
  background-color: var(--primaryDark);
}

.secondary {
  background-color: var(--secondary);
  border-color: var(--secondaryDark);
  color: #ffffff;
  outline-color: var(--secondaryDark);
}

.secondary:hover {
  background-color: var(--secondaryLight);
}

.secondary:active:not:hover {
  background-color: var(--secondaryDark);
}

.gray {
  background-color: var(--gray);
  border-color: var(--grayDark);
  color: #ffffff;
  outline-color: var(--grayDark);
}

.gray:hover {
  background-color: var(--grayLight);
}

.gray:active:not:hover {
  background-color: var(--grayDark);
}
