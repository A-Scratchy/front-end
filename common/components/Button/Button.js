import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Link from 'next/link';
import OutboundLink from 'common/components/OutboundLink/OutboundLink';
import styles from './Button.css';

Button.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  fullWidth: PropTypes.bool,
  href: PropTypes.string,
  onClick: PropTypes.func,
  tabIndex: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  theme: PropTypes.oneOf(['primary', 'secondary', 'slate']),
};

Button.defaultProps = {
  children: undefined,
  className: '',
  disabled: false,
  fullWidth: false,
  href: undefined,
  onClick: undefined,
  tabIndex: 0,
  theme: 'primary',
};

function Button({
  className, children, disabled, fullWidth, href, onClick, tabIndex, theme,
}) {
  // TODO: Handle non-string input for analytics event label on both outbound and scroll link
  // Example: SVG as a child

  const buttonClassNames = classNames(styles.Button, className, styles[theme], {
    [styles.disabled]: disabled,
    [styles.fullWidth]: fullWidth,
  });

  // MARK: Non-navigational button is checked first as it is the most common implementation
  if (!href) {
    return (
      <button
        className={buttonClassNames}
        disabled={disabled}
        onClick={disabled ? () => {} : onClick}
        tabIndex={tabIndex}
      >
        {children}
      </button>
    );
  }

  // MARK: Outbound link (leaves OperationCode host)
  if (href.includes('http')) {
    return (
      <OutboundLink
        analyticsEventLabel={children}
        className={classNames(buttonClassNames, styles.outboundLink)}
        hasIcon
        href={href}
        onClick={onClick}
        tabIndex={tabIndex}
      >
        {children}
      </OutboundLink>
    );
  }

  // MARK: Internal navigation link button
  return (
    <Link
      href={href}
      prefetch
    >
      {/* eslint-disable */}
      {/* TODO: resolve a bunch of a11y rules... */}
      <a className={buttonClassNames} onClick={onClick}>
        {children}
      </a>
      {/* eslint-enable */}
    </Link>
  );
}

export default Button;
