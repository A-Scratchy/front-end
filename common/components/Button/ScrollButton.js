import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import ReactGA from 'react-ga';
import { Link as ScrollLink, Events as ScrollEvent } from 'react-scroll';
import styles from './Button.css';

ScrollButton.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  fullWidth: PropTypes.bool,
  href: PropTypes.string,
  onClick: PropTypes.func,
  tabIndex: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  theme: PropTypes.oneOf(['primary', 'secondary', 'slate']),
};

ScrollButton.defaultProps = {
  className: '',
  fullWidth: false,
  href: '',
  onClick: () => {},
  tabIndex: 0,
  theme: 'primary',
};

function ScrollButton({
 className, children, fullWidth, href, onClick, tabIndex, theme
}) {
  const buttonClassNames = classNames(
styles.Button, className, {
    [styles.primary]: theme === 'primary',
    [styles.secondary]: theme === 'secondary',
    [styles.slate]: theme === 'slate',
    [styles.fullWidth]: fullWidth,
  }
);

  // TODO: Handle non-string input for analytics event label on both outbound and scroll link
  // Example: SVG as a child
  const eventDetails = {
    category: 'Scroll Button Clicked',
    action: `[${children}] from ${window.location.pathname}`,
  };

  const isProd = process.env.NODE_ENV === 'production';

  // Report scroll link button clicks to Google Analytics
  if (isProd) {
    ScrollEvent.scrollEvent.register('begin', () => {
      ReactGA.event(eventDetails);
    });
  }

  const clickHandler = () => {
    if (!isProd) {
      const analyticsMessage = `Analytics disabled.
        Message: ${eventDetails.category} - ${eventDetails.action}`;

      // eslint-disable-next-line no-console
      console.log(analyticsMessage);
      onClick(analyticsMessage);
      return;
    }

    onClick();
  };

  const onEnterHandler = ev => (ev.key === 'Enter' ? clickHandler : () => {});

  return (
    <ScrollLink
      className={buttonClassNames}
      duration={400}
      onClick={clickHandler}
      onKeyDown={onEnterHandler}
      smooth
      tabIndex={tabIndex}
      to={href}
    >
      {children}
    </ScrollLink>
  );
}

export default ScrollButton;
