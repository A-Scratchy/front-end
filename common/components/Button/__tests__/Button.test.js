/* eslint-env jest */
import React from 'react';
import { mount } from 'enzyme';
import createSnapshotTest from 'test-utils/createSnapshotTest';

import Button from '../Button';

describe('Button', () => {
  test('should render with just required props passed', () => {
    createSnapshotTest(<Button>Test</Button>);
  });

  test('should render properly with some props assigned', () => {
    createSnapshotTest(
      <Button
        theme="secondary"
        data-custom-attr="custom stuff here"
        disabled
        fullWidth
        type="submit"
      >
        Test
      </Button>,
    );
  });

  test('should render without a generated span when children is PropTypes.node', () => {
    const testText = 'Testing No Span';

    const ButtonInstance = mount(
      <Button>
        <b>{testText}</b>
      </Button>,
    );

    expect(
      ButtonInstance.containsAnyMatchingElements([
        <span>{testText}</span>,
        <span>
          <b>{testText}</b>
        </span>,
      ]),
    ).toEqual(false);
  });

  test('should render with a generated span when children is PropTypes.string', () => {
    const testText = 'Testing With Span';

    const ButtonInstance = mount(<Button>{testText}</Button>);

    expect(ButtonInstance.containsAllMatchingElements([<span>{testText}</span>])).toEqual(true);
  });
});
