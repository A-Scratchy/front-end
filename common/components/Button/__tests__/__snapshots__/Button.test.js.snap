// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render properly with some props assigned 1`] = `
<button
  className="Button secondary disabled fullWidth"
  disabled={true}
  onClick={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
  type="submit"
>
  Test
</button>
`;

exports[`Button should render with just required props passed 1`] = `
<button
  className="Button primary"
  disabled={false}
  onClick={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
  type="button"
>
  Test
</button>
`;
